File: Go/handlers/historical_data.go
package handlers

import (
    "github.com/gin-gonic/gin"
    "github.com/nats-io/nats.go"
    "net/http"
    "stock_broker/models"
    "stock_broker/services"
)

func GetHistoricalData(nc *nats.Conn) gin.HandlerFunc {
    return func(c *gin.Context) {
        var query models.HistoricalDataRequestMessage
        if err := c.ShouldBindQuery(&query); err != nil {
            c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request"})
            return
        }

        message, err := json.Marshal(query)
        if err != nil {
            c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to serialize historical data request"})
            return
        }

        err = services.PublishToNATS(nc, "historical_data_request", message)
        if err != nil {
            c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to publish request to NATS"})
            return
        }

        c.JSON(http.StatusOK, gin.H{"status": "success", "message": "Request for historical data published to NATS"})
    }
}

File: Rust/src/handlers/historical_data.rs
use actix_web::{web, HttpResponse, Responder};
use serde_json::json;
use nats::asynk::Connection;

extern create stock_broker_application;

use stock_broker_application::services::nats_service::publish_to_nats;
use stock_broker_application::models::HistoricalDataRequestMessage;


pub async fn get_historical_data(nats: web::Data<Connection>, query: web::Query<HistoricalDataRequestMessage>) -> impl Responder {
    let message_json = match serde_json::to_string(&query.into_inner()) {
        Ok(json) => json,
        Err(e) => {
            eprintln!("Serialization error: {}", e);
            return HttpResponse::InternalServerError().json(json!({
                "status": "error",
                "message": "Failed to serialize historical data request"
            }));
        }
    };

    match publish_to_nats(&nats, "historical_data_request", message_json.as_bytes()).await {
        Ok(_) => HttpResponse::Ok().json(json!({
            "status": "success",
            "message": "Request for historical data published to NATS"
        })),
        Err(e) => {
            eprintln!("NATS publish error: {}", e);
            HttpResponse::InternalServerError().json(json!({
                "status": "error",
                "message": "Failed to publish request to NATS"
            }))
        }
    }
}
